[Spam tokens] manually hide tokens

schmanu opened the pull request:
## What it solves
Unwanted / spam tokens could not be hidden from the interface.

Resolves #1337 

## How this PR fixes it
- New button in Assets page to manually hide / unhide tokens
- new parameter for `useBalances` to get all or only not hidden tokens

## How to test it
- Hide some tokens in the Assets page of a safe
- These tokens should be hidden for all Safes on that chain
  - Not appear in assets
  - not counted in total balance
  - not appear through safe apps sdk `getBalance`
  - not show in Dashboard Overview

## Analytics changes
- New events for hiding / unhiding tokens
## Screenshots

katspaugh commented on src/hooks/useBalances.ts:
```diff
@@ -3,22 +3,52 @@ import isEqual from 'lodash/isEqual'
+import { BigNumber } from 'ethers'
 
-const useBalances = (): {
+const removeHiddenBalances = (
```
`remove` sounds like a destructive operation. `filterHiddenTokens` would be better.

katspaugh commented on src/hooks/useBalances.ts:
```diff
@@ -3,22 +3,52 @@ import isEqual from 'lodash/isEqual'
+  balances: SafeBalanceResponse,
+  hiddenAssets: HiddenAssetsOnChain,
+): SafeBalanceResponse => {
+  const fiatTotalVisible = safeFormatUnits(
```
I wouldn't mess with the total. Hidden doesn't mean it's not there at all. I would keep the total as is sent from the backend.

schmanu replied:
During problem shaping we made a list of all places where hiding an asset could have influence and discussed each.

For total balances we thought it would be nice as a Safe could receive scam tokens with a value denoted to them. 
Also using different token lists and manually hiding tokens would have consistent effects like this as the total balance from our tx service would also change if we filter by default list.

katspaugh replied:
When the filter is on the backend, it obviously should change the total. Tokens hidden on the client-side affecting the total makes the client-side logic too complex. We should avoid messing with the financially significant data on the client side. Besides, it creates more divergence between mobile and desktop, both technically and product-wise.
On mobile, the balance would still include the hidden tokens and it might confuse the user.

katspaugh commented on src/store/hiddenAssetsSlice.ts:
```diff
@@ -0,0 +1,54 @@
+import type { RootState } from '.'
+
+export type HiddenAssetsOnChain = {
+  [tokenAddress: string]: string
```
Why does it need to be a map instead of just a flat list of token addresses?

katspaugh commented on src/store/hiddenAssetsSlice.ts:
```diff
@@ -0,0 +1,54 @@
+import { createSelector, createSlice, type PayloadAction } from '@reduxjs/toolkit'
+import type { RootState } from '.'
+
+export type HiddenAssetsOnChain = {
```
What does "onchain" mean here?

katspaugh commented on src/components/balances/HiddenAssetsProvider/index.tsx:
```diff
@@ -0,0 +1,119 @@
+  visibleAssets: [],
+})
+
+const HiddenAssetsProvider = ({ children }: { children: ReactNode }): ReactElement => {
```
Why do you need a provider when we usually just use selectors and dispatchers directly? Feels like an unnecessary abstraction layer.

katspaugh reviewed (commented):
I would suggest rethinking the terminology. E.g. UserTokenBlocklist instead of HiddenAssets.

katspaugh commented on src/components/balances/AssetsTable/index.tsx:
```diff
@@ -12,118 +13,258 @@ import useIsGranted from '@/hooks/useIsGranted'
+      showHiddenAssets
+        ? balances.items
+        : balances.items?.filter(
+            (item) => item.tokenInfo.type === TokenType.NATIVE_TOKEN || !hiddenAssets.includes(item.tokenInfo.address),
```
Why is the native token filtered out?

schmanu replied:
Currently because I can't imagine people actually want to hide it on purpose and because the native token has no address. 
So we would need some hacky workaround such as "If the Zero address is in the hidden tokens, we do not show the native token"

katspaugh replied:
I don't see how it's a reason not to allow the user to do it if they want to. Zero address (0x0000...) is what backend sends us the native token's address IIRC.

schmanu replied:
Okay I'll remove this condition and the disabled flags in the components.
I checked and true: Zero address is returned by backend.

katspaugh reviewed (commented):
As you TODO says, there's a lot of new logic inside the AssetsTable that would be good to extract into a module dedicated to token hiding.

schmanu commented on src/components/balances/AssetsTable/index.tsx:
```diff
@@ -12,118 +13,258 @@ import useIsGranted from '@/hooks/useIsGranted'
+  const [hidingAsset, setHidingAsset] = useState<string>()
 
-  const shouldHideActions = !isGranted
+  // TODO: move logic into hook
```
I just finished this TODO :) Now it is ready.

schmanu commented on src/components/balances/AssetsTable/index.tsx:
```diff
@@ -12,118 +13,258 @@ import useIsGranted from '@/hooks/useIsGranted'
+                  </Track>
+                )}
+                {showHiddenAssets ? (
+                  <Track {...ASSETS_EVENTS.HIDE_CHECKBOX}>
```
I will move the tracking to all dialog buttons. Makes more sense 👍

katspaugh commented on src/hooks/useBalances.ts:
```diff
@@ -3,22 +3,52 @@ import isEqual from 'lodash/isEqual'
+  const fiatTotalVisible = safeFormatUnits(
+    balances.items
+      .reduce((acc, balanceItem) => {
+        if (typeof hiddenAssets?.[balanceItem.tokenInfo.address] !== 'undefined') {
```
I still think we shouldn't touch the total balance. What did backend and mobile say to this?

Sidenote: we use `typeof === 'undefined'` for entirely undeclared variables, for regular null/undefined-values you can simply do `== null` or `=== undefined`.

schmanu replied:
This is the last discussion point where I want product to have the last call. If @johannesmoormann decides to leave it untouched it is a very quick change. 

The decision was made based on the fact that there are spam tokens which are part of liquidity pools leading to price oracles returning a USD value for them. This could cloud the real total USD balance / value of the Safe.

katspaugh replied:
I'm talking about the solution with summing up fiat values on the client side.
Moving this logic to `useVisibleBalances` as discussed on Slack makes it slightly better.

katspaugh commented on src/hooks/useVisibleBalances.ts:
```diff
@@ -0,0 +1,21 @@
+  const hiddenTokens = useHiddenTokens()
+  const { items, fiatTotal } = balances.balances
+
+  return {
+    ...balances,
+    balances: {
+      items: items.filter((item) => !hiddenTokens.includes(item.tokenInfo.address)),
+      fiatTotal,
+    },
```
Bear in mind this will filter the list and create a new object on each render. Would be good to memo it.

katspaugh commented on src/components/balances/AssetsTable/useHideAssets.ts:
```diff
@@ -0,0 +1,88 @@
+import { useCallback, useState } from 'react'
+
+// This is the default for MUI Collapse
+export const COLLAPSE_TIMEOUT_MS = 300
```
No need to export it, it's used in the same module.

schmanu replied:
The test is using it.

usame-algan commented:
There is some overlap/spacing issues on smaller screens. Do we have mobile designs? cc @liliiaorlenko 

<img width="488" alt="Screenshot 2023-01-05 at 11 30 37" src="https://user-images.githubusercontent.com/5880855/210759523-1aa2a57b-e9b5-46d7-8147-f46f42c89d3f.png">

katspaugh reviewed (approved):
Looks good from my side, apart from the remaining comments.

katspaugh commented on src/components/balances/AssetsTable/useHideAssets.ts:
```diff
@@ -0,0 +1,90 @@
+  const toggleAsset = useCallback(
+    (address: string) => {
+      if (assetsToHide.includes(address)) {
+        const newAssetsToHide = [...assetsToHide]
+        newAssetsToHide.splice(newAssetsToHide.indexOf(address), 1)
```
We usually use `.filter` for this.

katspaugh commented on src/components/balances/AssetsTable/useHideAssets.ts:
```diff
@@ -0,0 +1,90 @@
+
+      const assetIsHidden = hiddenAssets.includes(address)
+      if (!assetIsHidden) {
+        setAssetsToHide([...assetsToHide, address])
```
And for this, IMHO, concat is better. `setAssetsToHide((items) => items.concat(address))`.

francovenica commented:
1 - The feature works fine for desktop, but for mobile it not usable: you cannot see the checkboxes while the "hide tokens menu" is open so you cannot use the feature.
I'd suggest hiding the button for mobile views until we can design it to be mobile friendly
![image](https://user-images.githubusercontent.com/4503659/210922081-73634da4-3561-43c9-a06c-538815555e49.png)

2 - The widget that counts the amount of tokens in the homepage still shows all the tokens you have, hidden or not. I'd argue that if you want to hide tokens, you don't want them to be counted with the ones you didn't hide.
I'd propose not count them at all or, if you have hidden tokens, you show a number in "()" indicating how many of them are hidden ones. 
![image](https://user-images.githubusercontent.com/4503659/210922760-e1448dec-a6b4-4b9c-ae5e-d47af0e52533.png)

3 - Just refreshing a page with a bunch of tokens throws a ton of GTM calls. No tokens are hidden right now. 
I'd assume the calls should be:
When the menu is opened/closed
When the cancel/deselect button are clicked
When the save button is clicked, sending how many tokens are being hidden by that save action.
When the "eye" icon is clicked to hide a token
![image](https://user-images.githubusercontent.com/4503659/210923292-479551ef-e3b3-4a93-b42b-526789797024.png)

Questions/Concerns:
- Deselect all seems to clear the storage, so if you use it in a safe with barely no tokens you will end up deselecting a ton of tokens that you might have hidden in another safe without you knowing it. Maybe deselect can just trigger the click action in the checkboxes that are visible instead, so deselecting in one safe won't affect the others (unless they share the token being hidden/shown)

- The total safe balance is not being affected by hiding tokens. I've been told that we are going to ask Johannes first about this, but I personally think that yes, the balances should reflect only the visible tokens

- Is unlikely that somebody will set the "Safe" token as invisible, but if somebody does the token widget in the top bar and in the Home widget is still there. I'm gonna assume that is expected behaviour.

schmanu commented:
Hey @francovenica 

I addresses most issues:
1. I removed the button in mobile.
2. The widget now does not count hidden assets anymore.
3. I moved the tracking to how we track the totalBalance (useMetaEvents). That should fix these issues. Although I wasn't able to reproduce those issues :/ 

For the questions:
1. _Deselect all_ was implemented wrong. Now it only deselects tokens which are present in the opened Safe
2. After talking with product we agreed that the total balance should be adjusted. I added that now.
3. That is expected. The Safe token widget should not be affected by this.

francovenica commented:
Yes, all the issues were addressed. The GTM calls are also working fine (not 10 calls just by refreshing the page anymore)

usame-algan commented on src/components/balances/AssetsHeader/index.tsx:
```diff
@@ -1,20 +1,42 @@
-const AssetsHeader = ({ currencySelect = false }: { currencySelect?: boolean }): ReactElement => {
+import HiddenTokenButton from '../HiddenTokenButton'
+
+const AssetsHeader = ({
```
I think we can simplify this component by only passing `children?: ReactNode` instead of the other props since they are not being used in `AssetsHeader`. This would reduce the amount of nesting we have and we could include the child components directly in the balances page like this:

```
<AssetsHeader>
    <Box display="flex" flexDirection="row" alignItems="center" gap={1}>
      {hiddenAssets && (
        <HiddenTokenButton showHiddenAssets={showHiddenAssets} toggleShowHiddenAssets={toggleShowHiddenAssets} />
      )}
      <CurrencySelect />
    </Box>
</AssetsHeader>
```

usame-algan replied:
As discussed I will push this small refactor

katspaugh commented on src/hooks/useVisibleBalances.ts:
```diff
@@ -0,0 +1,59 @@
+  return currentPrecision < PRECISION ? balance : balance.slice(0, floatingPointPosition + PRECISION + 1)
+}
+
+const filterHiddenBalances = (balances: SafeBalanceResponse, hiddenAssets: string[]): SafeBalanceResponse => {
+  const fiatTotalVisible = safeFormatUnits(
+    balances.items
+      .reduce((acc, balanceItem) => {
+        if (hiddenAssets.includes(balanceItem.tokenInfo.address)) {
+          return acc.sub(safeParseUnits(truncateNumber(balanceItem.fiatBalance), PRECISION) || 0)
+        }
+        return acc
+      }, BigNumber.from(balances.fiatTotal === '' ? 0 : safeParseUnits(truncateNumber(balances.fiatTotal), PRECISION)))
+      .toString(),
+    PRECISION,
+  )
+
+  const balanceItemsVisible = balances.items.filter(
+    (balanceItem) => !hiddenAssets.includes(balanceItem.tokenInfo.address),
+  )
+
+  return { fiatTotal: fiatTotalVisible, items: balanceItemsVisible }
+}
```
Could you please split this into two functions, `filterHiddenTokens` and `getFilteredTotal`?

katspaugh reviewed (approved):
Looks good!

Btw according to the SafeBalanceResponse type, `items` and `fiatTotal` are the only fields in this object, so you don't need the `...balances`.

schmanu commented:
@katspaugh 
The names are a bit misleading there the returned object from `const balances = useBalances` will return
```ts
{
 balances: SafeBalanceResponse
  loading: boolean
  error?: string
}
```

So I spread to use the `loading` and `error` props from the `useBalances` hook as the visibleBalances would be loading / erroneous too.
